Mage/src/main/java/mage/abilities/effects/common/combat/CantAttackYouUnlessPayAllEffect.java

```diff
@@ -6,42 +6,66 @@
 public class CantAttackYouUnlessPayAllEffect extends PayCostToAttackBlockEffectImpl {
 
+    public enum Scope {
+        YOU_AND_CONTROLLED_PLANESWALKERS(true, true,
+            "you or planeswalkers you control", "of those creatures"),
+        YOU_ONLY(true, false,
+            "you", "creature they control that's attacking you"),
+        CONTROLLED_PLANESWALKERS_ONLY(false, true,
+            "planeswalkers you control", "creature they control that's attacking a planeswalker you control");
+
+        final boolean attackingYou;
+        final boolean attackingControlledPlaneswalkers;
+        // text replacing [!] in "[...] can't attack [!] unless [...] pays ..."
+        final String firstText;
+        // text replacing [!] in "unless their controller pays [...] for each [!]"
+        final String secondText;
+
+        Scope(boolean attackingYou, boolean attackingControlledPlaneswalkers, String firstText, String secondText) {
+            this.attackingYou = attackingYou;
+            this.attackingControlledPlaneswalkers = attackingControlledPlaneswalkers;
+            this.firstText = firstText;
+            this.secondText = secondText;
+        }
+    }
+
     private final FilterCreaturePermanent filterCreaturePermanent;
-    private final boolean payAlsoForAttackingPlaneswalker;
+    private final Scope scope; // Describe which attacked objects this effect cares about.
 
-    public CantAttackYouUnlessPayAllEffect(Cost cost) {
-        this(cost, false);
+    public CantAttackYouUnlessPayAllEffect(Duration duration, Cost cost) {
+        this(duration, cost, Scope.YOU_ONLY);
     }
 
-    public CantAttackYouUnlessPayAllEffect(Cost cost, boolean payAlsoForAttackingPlaneswalker) {
-        this(cost, payAlsoForAttackingPlaneswalker, null);
+    public CantAttackYouUnlessPayAllEffect(Duration duration, Cost cost, Scope scope) {
+        this(duration, cost, scope, StaticFilters.FILTER_PERMANENT_CREATURES);
     }
 
-    public CantAttackYouUnlessPayAllEffect(Cost cost, boolean payAlsoForAttackingPlaneswalker, FilterCreaturePermanent filter) {
-        super(Duration.WhileOnBattlefield, Outcome.Detriment, RestrictType.ATTACK, cost);
-        this.payAlsoForAttackingPlaneswalker = payAlsoForAttackingPlaneswalker;
+    public CantAttackYouUnlessPayAllEffect(Duration duration, Cost cost, Scope scope, FilterCreaturePermanent filter) {
+        super(duration, Outcome.Detriment, RestrictType.ATTACK, cost);
+        this.scope = scope;
         this.filterCreaturePermanent = filter;
-        staticText = (filterCreaturePermanent == null ? "Creatures" : filterCreaturePermanent.getMessage())
-                + " can't attack you "
-                + (payAlsoForAttackingPlaneswalker ? "or a planeswalker you control " : "")
-                + "unless their controller pays "
-                + (cost == null ? "" : cost.getText())
-                + " for each creature they control that's attacking you";
+        staticText = duration.toString()
```

Comment:
I think putting duration directly isn't sufficient and you need something like:
```
(duration.toString().isEmpty() ? "" : "duration.toString()" + ", ")
```

Btw, personally I find double indent (8 spaces) clearer when breaking a statement over multiple lines, not sure why your IDE is switching it to single indent (4 spaces).

Reply:
(I'll check the line break, I did tweak automatic settings recently.)